app/api/groups/get_group_progress.go

```diff
@@ -0,0 +1,175 @@
+	if !found {
+		return service.InsufficientAccessRightsError
+	}
+
+	itemsVisibleToUserSubquery := srv.Store.GroupItems().
+		Select(
+			"idItem, MIN(sCachedFullAccessDate) <= NOW() AS fullAccess, "+
+				"MIN(sCachedPartialAccessDate) <= NOW() AS partialAccess, "+
+				"MIN(sCachedGrayedAccessDate) <= NOW() AS grayedAccess").
+		Joins(`
+			JOIN (SELECT * FROM groups_ancestors WHERE (groups_ancestors.idGroupChild = ?)) AS ancestors
+			ON ancestors.idGroupAncestor = groups_items.idGroup`, userSelfGroupID).
+		Group("groups_items.idItem").
+		Having("fullAccess > 0 OR partialAccess > 0 OR grayedAccess > 0").SubQuery()
+
+	// Preselect item IDs since we want to use them twice (for end members stats and for final stats)
+	// There should not be many of them
+	var itemIDs []int64
+	service.MustNotBeError(srv.Store.ItemItems().Where("idItemParent IN (?)", itemParentIDs).
+		Joins("JOIN ? AS visible ON visible.idItem = items_items.idItemChild", itemsVisibleToUserSubquery).
+		Pluck("items_items.idItemChild", &itemIDs).Error())
+
+	// Preselect IDs of groups for that we will calculate the final stats.
+	// All the "end members" are descendants of these groups.
+	// There should not be too many of groups because we paginate on them.
+	var ancestorGroupIDs []int64
+	ancestorGroupIDQuery := srv.Store.GroupGroups().
+		Where("groups_groups.idGroupParent = ?", groupID).
+		Joins(`
+			JOIN groups AS group_child
+			ON group_child.ID = groups_groups.idGroupChild AND group_child.sType NOT IN('Team', 'UserSelf')`)
+	ancestorGroupIDQuery, apiError := service.ApplySortingAndPaging(r, ancestorGroupIDQuery, map[string]*service.FieldSortingParams{
+		"id": {ColumnName: "group_child.ID", FieldType: "int64"},
+	}, "id")
+	if apiError != service.NoError {
+		return apiError
+	}
+	ancestorGroupIDQuery = service.SetQueryLimit(r, ancestorGroupIDQuery, 10, 20)
+	service.MustNotBeError(ancestorGroupIDQuery.
+		Pluck("group_child.ID", &ancestorGroupIDs).Error())
+
+	endMembersStats := srv.Store.
+		Table("groups_attempts AS main_attempts FORCE INDEX(GroupItemMinusScoreBestAnswerDateID)").
+		Select(`
+			main_attempts.idItem AS idItem,
+			main_attempts.idGroup AS idGroup,
+			IFNULL(MAX(main_attempts.iScore), 0) AS iMaxEndMemberScore,
+			IFNULL(MAX(main_attempts.bValidated), 0) AS bMaxEndMemberValidated,
+			IFNULL(attempt_with_best_score.nbHintsCached, 0) AS nbEndMemberBestScoreHintsCached,
+			IFNULL(attempt_with_best_score.nbSubmissionsAttempts, 0) AS nbEndMemberBestScoreSubmissionAttempts,
+			IF(attempt_with_best_score.idGroup IS NULL,
+				0,
+				IF(MAX(main_attempts.bValidated),
+					TIMESTAMPDIFF(SECOND, MIN(main_attempts.sStartDate), MIN(main_attempts.sValidationDate)),
+					TIMESTAMPDIFF(SECOND, MIN(main_attempts.sStartDate), NOW())
+				)
+			) AS sEndMemberTimeSpent`).
+		Joins(`
+			JOIN groups_ancestors
+			ON groups_ancestors.idGroupChild = main_attempts.idGroup AND
+				groups_ancestors.idGroupChild != groups_ancestors.idGroupAncestor AND
```

Comment:
could user `bIsSelf`

Reply:
I'm not sure why but `EXPLAIN` says that using `bIsSelf` would make things twice worse (note 559796 becomes 1119589):
```
mysql> explain SELECT groups_ancestors.idGroupAncestor AS idGroup, items.ID AS idItem, AVG(IFNULL(attempt_with_best_score.iScore, 0)) AS iAverageScore, AVG(IFNULL(attempt_with_best_score.bValidated, 0)) AS iValidationRate, AVG(IFNULL(attempt_with_best_score.nbHintsCached, 0)) AS iAvgHintsRequested, AVG(IFNULL(attempt_with_best_score.nbSubmissionsAttempts, 0)) AS iAvgSubmissionsAttempts, AVG(IF(attempt_with_best_score.idGroup IS NULL, 0, ( SELECT IF(MAX(bValidated), TIMESTAMPDIFF(SECOND, MIN(sStartDate), MIN(sValidationDate)), TIMESTAMPDIFF(SECOND, MIN(sStartDate), NOW()) ) FROM groups_attempts WHERE idGroup = end_member.ID AND idItem = items.ID ) )) AS sAvgTimeSpent FROM `groups_ancestors` JOIN groups AS end_member ON end_member.ID = groups_ancestors.idGroupChild AND end_member.sType IN ('UserSelf', 'Team') JOIN (SELECT 1 as bKeepUser) AS keep_user ON end_member.sType = 'Team' OR ( SELECT 1 FROM groups_groups JOIN groups ON groups.ID = groups_groups.idGroupParent AND groups.sType != 'Team' WHERE idGroupChild = end_member.ID AND groups_groups.sType IN('invitationAccepted','requestAccepted','direct') AND idGroupParent IN ( SELECT idGroupChild FROM groups_ancestors AS ga WHERE ga.idGroupAncestor = groups_ancestors.idGroupAncestor AND NOT ga.bIsSelf ) LIMIT 1 ) = 1 JOIN items ON items.ID IN (211,212,213,214,215,216,217,218,219,221,222,223,224,225,226,227,228,229,311,312,313,314,315,316,317,318,319) LEFT JOIN groups_attempts AS attempt_with_best_score ON attempt_with_best_score.ID = ( SELECT ID FROM groups_attempts WHERE idGroup = end_member.ID AND idItem = items.ID ORDER BY idGroup, idItem, iMinusScore, sBestAnswerDate LIMIT 1 ) WHERE (groups_ancestors.idGroupChild != groups_ancestors.idGroupAncestor) AND (groups_ancestors.idGroupAncestor IN(11)) GROUP BY groups_ancestors.idGroupAncestor, items.ID ORDER BY groups_ancestors.idGroupAncestor, items.ID;
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+--------+-----------------------------------------------------------------+
| id | select_type        | table                   | type   | possible_keys                                        | key                                 | key_len | ref                                                                                | rows   | Extra                                                           |
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+--------+-----------------------------------------------------------------+
|  1 | PRIMARY            | <derived3>              | system | NULL                                                 | NULL                                | NULL    | NULL                                                                               |      1 | Using temporary; Using filesort                                 |
|  1 | PRIMARY            | items                   | index  | PRIMARY                                              | iVersion                            | 8       | NULL                                                                               |     43 | Using where; Using index                                        |
|  1 | PRIMARY            | end_member              | range  | PRIMARY,IDType                                       | IDType                              | 8       | NULL                                                                               | 559796 | Using where; Using index; Using join buffer (Block Nested Loop) |
|  1 | PRIMARY            | groups_ancestors        | eq_ref | idGroupAncestor,ancestor,descendant                  | idGroupAncestor                     | 16      | const,algorea_db.end_member.ID                                                     |      1 | Using where; Using index                                        |
|  1 | PRIMARY            | attempt_with_best_score | eq_ref | PRIMARY                                              | PRIMARY                             | 8       | func                                                                               |      1 | Using where                                                     |
|  6 | DEPENDENT SUBQUERY | groups_attempts         | ref    | idItem,GroupItem,GroupItemMinusScoreBestAnswerDateID | GroupItemMinusScoreBestAnswerDateID | 16      | algorea_db.end_member.ID,algorea_db.items.ID                                       |      1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | groups_groups           | ref    | parentchild,idGroupChild,idGroupParent               | idGroupChild                        | 8       | algorea_db.end_member.ID                                                           |      1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | groups                  | eq_ref | PRIMARY,IDType                                       | PRIMARY                             | 8       | algorea_db.groups_groups.idGroupParent                                             |      1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | ga                      | eq_ref | idGroupAncestor,ancestor,descendant                  | idGroupAncestor                     | 16      | algorea_db.groups_ancestors.idGroupAncestor,algorea_db.groups_groups.idGroupParent |      1 | Using where                                                     |
|  3 | DERIVED            | NULL                    | NULL   | NULL                                                 | NULL                                | NULL    | NULL                                                                               |   NULL | No tables used                                                  |
|  2 | DEPENDENT SUBQUERY | groups_attempts         | ref    | idItem,GroupItem,GroupItemMinusScoreBestAnswerDateID | GroupItem                           | 16      | algorea_db.end_member.ID,algorea_db.items.ID                                       |      1 | NULL                                                            |
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+--------+-----------------------------------------------------------------+
11 rows in set (0.00 sec)

mysql> explain SELECT groups_ancestors.idGroupAncestor AS idGroup, items.ID AS idItem, AVG(IFNULL(attempt_with_best_score.iScore, 0)) AS iAverageScore, AVG(IFNULL(attempt_with_best_score.bValidated, 0)) AS iValidationRate, AVG(IFNULL(attempt_with_best_score.nbHintsCached, 0)) AS iAvgHintsRequested, AVG(IFNULL(attempt_with_best_score.nbSubmissionsAttempts, 0)) AS iAvgSubmissionsAttempts, AVG(IF(attempt_with_best_score.idGroup IS NULL, 0, ( SELECT IF(MAX(bValidated), TIMESTAMPDIFF(SECOND, MIN(sStartDate), MIN(sValidationDate)), TIMESTAMPDIFF(SECOND, MIN(sStartDate), NOW()) ) FROM groups_attempts WHERE idGroup = end_member.ID AND idItem = items.ID ) )) AS sAvgTimeSpent FROM `groups_ancestors` JOIN groups AS end_member ON end_member.ID = groups_ancestors.idGroupChild AND end_member.sType IN ('UserSelf', 'Team') JOIN (SELECT 1 as bKeepUser) AS keep_user ON end_member.sType = 'Team' OR ( SELECT 1 FROM groups_groups JOIN groups ON groups.ID = groups_groups.idGroupParent AND groups.sType != 'Team' WHERE idGroupChild = end_member.ID AND groups_groups.sType IN('invitationAccepted','requestAccepted','direct') AND idGroupParent IN ( SELECT idGroupChild FROM groups_ancestors AS ga WHERE ga.idGroupAncestor = groups_ancestors.idGroupAncestor AND NOT ga.bIsSelf ) LIMIT 1 ) = 1 JOIN items ON items.ID IN (211,212,213,214,215,216,217,218,219,221,222,223,224,225,226,227,228,229,311,312,313,314,315,316,317,318,319) LEFT JOIN groups_attempts AS attempt_with_best_score ON attempt_with_best_score.ID = ( SELECT ID FROM groups_attempts WHERE idGroup = end_member.ID AND idItem = items.ID ORDER BY idGroup, idItem, iMinusScore, sBestAnswerDate LIMIT 1 ) WHERE (NOT groups_ancestors.bIsSelf) AND (groups_ancestors.idGroupAncestor IN (11)) GROUP BY groups_ancestors.idGroupAncestor, items.ID ORDER BY groups_ancestors.idGroupAncestor, items.ID;
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+---------+-----------------------------------------------------------------+
| id | select_type        | table                   | type   | possible_keys                                        | key                                 | key_len | ref                                                                                | rows    | Extra                                                           |
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+---------+-----------------------------------------------------------------+
|  1 | PRIMARY            | <derived3>              | system | NULL                                                 | NULL                                | NULL    | NULL                                                                               |       1 | Using temporary; Using filesort                                 |
|  1 | PRIMARY            | items                   | index  | PRIMARY                                              | iVersion                            | 8       | NULL                                                                               |      43 | Using where; Using index                                        |
|  1 | PRIMARY            | end_member              | index  | PRIMARY,IDType                                       | IDType                              | 9       | NULL                                                                               | 1119589 | Using where; Using index; Using join buffer (Block Nested Loop) |
|  1 | PRIMARY            | groups_ancestors        | eq_ref | idGroupAncestor,ancestor,descendant                  | idGroupAncestor                     | 16      | const,algorea_db.end_member.ID                                                     |       1 | Using where                                                     |
|  1 | PRIMARY            | attempt_with_best_score | eq_ref | PRIMARY                                              | PRIMARY                             | 8       | func                                                                               |       1 | Using where                                                     |
|  6 | DEPENDENT SUBQUERY | groups_attempts         | ref    | idItem,GroupItem,GroupItemMinusScoreBestAnswerDateID | GroupItemMinusScoreBestAnswerDateID | 16      | algorea_db.end_member.ID,algorea_db.items.ID                                       |       1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | groups_groups           | ref    | parentchild,idGroupChild,idGroupParent               | idGroupChild                        | 8       | algorea_db.end_member.ID                                                           |       1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | groups                  | eq_ref | PRIMARY,IDType                                       | PRIMARY                             | 8       | algorea_db.groups_groups.idGroupParent                                             |       1 | Using where                                                     |
|  4 | DEPENDENT SUBQUERY | ga                      | eq_ref | idGroupAncestor,ancestor,descendant                  | idGroupAncestor                     | 16      | algorea_db.groups_ancestors.idGroupAncestor,algorea_db.groups_groups.idGroupParent |       1 | Using where                                                     |
|  3 | DERIVED            | NULL                    | NULL   | NULL                                                 | NULL                                | NULL    | NULL                                                                               |    NULL | No tables used                                                  |
|  2 | DEPENDENT SUBQUERY | groups_attempts         | ref    | idItem,GroupItem,GroupItemMinusScoreBestAnswerDateID | GroupItem                           | 16      | algorea_db.end_member.ID,algorea_db.items.ID                                       |       1 | NULL                                                            |
+----+--------------------+-------------------------+--------+------------------------------------------------------+-------------------------------------+---------+------------------------------------------------------------------------------------+---------+-----------------------------------------------------------------+
11 rows in set (0.00 sec)```